Add conventions for CLI Spans (#1588)
open-telemetry/semantic-conventions: docs/cli/cli-spans.md

Reviewer: Is this related to the following OTEP?
* https://github.com/open-telemetry/oteps/pull/258
  Following this OTEP the [opentelemetry-specification](https://github.com/open-telemetry/opentelemetry-specification) will be updated to define span propagation when launching CLI programs.

CC @adrielp

Author: Nope

Author: > @xrmx could you please create an issue to track this future work and maybe add a TODO comment in the markdown saying that context propagation will be added in the future? Thanks!

I'm not involved with the OTEP linked above, not sure I'm the right person to track the followups :)

Author: Added

<!--- Hugo front matter used to generate the website version of this page:
linkTitle: CLI
--->

# Semantic Conventions for CLI (Command Line Interface) programs

**Status**: [Experimental][DocumentStatus]

This document defines semantic conventions to apply when instrumenting CLI programs, both as a caller and as callee. This document is intended for short-lived programs that end their execution, i.e. not daemon or long running background tasks.

Span kind SHOULD be `INTERNAL` when the traced program is the callee or `CLIENT` when the caller is tracing another program.

The span name SHOULD be set to `{process.executable.name}`.
Instrumentations that have additional context about executed commands MAY use a different low-cardinality span name format and SHOULD document it.

Span status SHOULD be set to `Error` if `{process.exit.code}` is not 0.

<!-- TODO: context propagation https://github.com/open-telemetry/semantic-conventions/issues/1612 -->

> 

Reviewer 2: ```suggestion

# Execution (callee) spans

```

maybe?

|---|---|---|---|---|---|
| [`process.executable.name`](/docs/attributes-registry/process.md) | string | The name of the process executable. On Linux based systems, can be set to the `Name` in `proc/[pid]/status`. On Windows, can be set to the base name of `GetProcessImageFileNameW`. | `otelcol` | `Required` | ![Experimental](https://img.shields.io/badge/-experimental-blue) |
| [`process.exit.code`](/docs/attributes-registry/process.md) | int | The exit code of the process. | `127` | `Required` | ![Experimental](https://img.shields.io/badge/-experimental-blue) |
| [`process.pid`](/docs/attributes-registry/process.md) | int | Process identifier (PID). | `1234` | `Required` | ![Experimental](https://img.shields.io/badge/-experimental-blue) |
| [`error.type`](/docs/attributes-registry/error.md) | string | Describes a class of error the operation ended with. [1] | `timeout`; `java.net.UnknownHostException`; `server_certificate_invalid`; `500` | `Conditionally Required` if and only if process.exit.code is not 0 | ![Stable](https://img.shields.io/badge/-stable-lightgreen) |

> | [`process.command_args`](/docs/attributes-registry/process.md) | string[] | All the command arguments (including the command/executable itself) as received by the process. On Linux-based systems (and some other Unixoid systems supporting procfs), can be set according to the list of null-delimited strings extracted from `proc/[pid]/cmdline`. For libc-based executables, this would be the full argv vector passed to `main`. | `["cmd/otecol", "--config=config.yaml"]` | `Recommended` | ![Experimental](https://img.shields.io/badge/-experimental-blue) |

Reviewer 3: Do you think we should comment on anything about not including/redacting sensible content?

Author: This is a sensible concern, but since this text comes from the process model I think we could revise the text in another PR.

Author: Opened a followup issues here https://github.com/open-telemetry/semantic-conventions/issues/1620

The cardinality of `error.type` within one instrumentation library SHOULD be low.
Telemetry consumers that aggregate data from multiple instrumentation libraries and applications
should be prepared for `error.type` to have high cardinality at query time when no
additional filters are applied.

If the operation has completed successfully, instrumentations SHOULD NOT set `error.type`.

If a specific domain defines its own set of error identifiers (such as HTTP or gRPC status codes),
it's RECOMMENDED to:

- Use a domain-specific attribute
- Set `error.type` to capture all errors, regardless of whether they are defined within the domain-specific set or not.

---

`error.type` has the following list of well-known values. If one of them applies, then the respective value MUST be used; otherwise, a custom value MAY be used.

| Value  | Description | Stability |
|---|---|---|
| `_OTHER` | A fallback error value to be used when the instrumentation doesn't define a custom value. | ![Stable](https://img.shields.io/badge/-stable-lightgreen) |

<!-- markdownlint-restore -->
<!-- prettier-ignore-end -->
<!-- END AUTOGENERATED TEXT -->
<!-- endsemconv -->

> 

Reviewer 2: ```suggestion

# Client (caller) spans

```

<!--- Hugo front matter used to generate the website version of this page:
linkTitle: CLI
--->

# Semantic Conventions for CLI (Command Line Interface) programs

**Status**: [Experimental][DocumentStatus]

> This document defines semantic conventions to apply when instrumenting CLI programs.
> operations to attributes on a Span.

Reviewer 4: Nit typo

```suggestion
This document defines semantic conventions to apply when instrumenting CLI programs'
operations to attributes on a Span.
```

Author: I think the second line was a copy and paste error, reworked the text a bit.

The **span name** SHOULD be of the format `{process.executable.name}`. In case this is

> not informative enough because of interpreters calling other files, e.g. `python script.py` or in

Reviewer 2: how would generic instrumentation know whether an arbitrary string is informative or not?

Reviewer 5: The history of this was not generic, rather in context of click which would have a better value I think. How might we clarify this better? https://github.com/open-telemetry/semantic-conventions/issues/1577

Author: Yep, the context of this is building an instrumentation for a library (click) used for building CLIs. So there are cases where we have more clue :)

cases of subcommands, e.g. `flask run mycommand` instrumentation MAY provide a more

> descriptive span name of the format `process.command_line` if appropriate.

Author: ```suggestion
descriptive span name of the format `{process.command_line}` if appropriate.
```

> **Warning**

> > The `process.command_line` value can have high cardinality.

Author: ```suggestion
> The `process.command_line` value can have high cardinality in case of parameters passed by the user. It is recommended to strip parameters if there is enough domain knowledge for being able to do so.
```

Reviewer 4: Good call, I think this is important. Could this be added to table below and as an update to the original `process` [registry](https://github.com/open-telemetry/semantic-conventions/blob/25e0baecd6393d10c1b87a7074520985f0da2fac/model/process/registry.yaml#L87-L95)?

Reviewer 5: should this doc link to the process registry or is it assumed? personally, I didn't know there was one until you mentioned it..

Reviewer 5: actually it is mentioned below, possibly handy to say up front that "a CLI span is defined in terms of process attributes" with a link to that?

> 

Reviewer 2: let's mention that span status should be set to error when the exit code is not `0`

Author: Done

<!-- semconv span.cli -->
<!-- NOTE: THIS TEXT IS AUTOGENERATED. DO NOT EDIT BY HAND. -->
<!-- see templates/registry/markdown/snippet.md.j2 -->
<!-- prettier-ignore-start -->
<!-- markdownlint-capture -->
<!-- markdownlint-disable -->

| Attribute  | Type | Description  | Examples  | [Requirement Level](https://opentelemetry.io/docs/specs/semconv/general/attribute-requirement-level/) | Stability |
|---|---|---|---|---|---|
| [`process.command_args`](/docs/attributes-registry/process.md) | string[] | All the command arguments (including the command/executable itself) as received by the process. On Linux-based systems (and some other Unixoid systems supporting procfs), can be set according to the list of null-delimited strings extracted from `proc/[pid]/cmdline`. For libc-based executables, this would be the full argv vector passed to `main`. | `["cmd/otecol", "--config=config.yaml"]` | `Recommended` | ![Experimental](https://img.shields.io/badge/-experimental-blue) |

> | [`process.command_line`](/docs/attributes-registry/process.md) | string | The full command used to launch the process as a single string representing the full command. On Windows, can be set to the result of `GetCommandLineW`. Do not set this if you have to assemble it just for monitoring; use `process.command_args` instead. | `C:\cmd\otecol --config="my directory\config.yaml"` | `Recommended` | ![Experimental](https://img.shields.io/badge/-experimental-blue) |

Reviewer 2: the command line and command args are the same, but in a different format. Why specify both?

Author: Removed the stringified one, most languages follow C argv like interface.

<!--- Hugo front matter used to generate the website version of this page:
linkTitle: CLI
--->

# Semantic Conventions for CLI (Command Line Interface) programs

**Status**: [Experimental][DocumentStatus]

This document defines semantic conventions to apply when instrumenting CLI programs.

> A CLI span is defined in terms of [process attributes][ProcessRegistry].

Reviewer 2: I don't understand what this sentence means. Is it necessary?

Author: I'll drop it since all the attributes in the table link the the process registry already.

> The **span name** SHOULD be of the format `{process.executable.name}`. In case this is
> not informative enough because of interpreters calling other files, e.g. `python script.py` or in
> cases of subcommands, e.g. `flask run mycommand` instrumentation MAY provide a more
> descriptive span name providing a stringified version of `{process.command_args}` if appropriate.
> 
> > **Warning**
> > The `process.command_args` value can have high cardinality in case of parameters passed by the user.
> > It is recommended to strip parameters if there is enough domain knowledge for being able to do so.

Reviewer 2: I'd rephrase it to

```suggestion
The span name SHOULD, when possible, match low-cardinality representation of command line. All variable parameters SHOULD be replaced with a `?` placeholder.

If instrumentation does not parse and sanitize command line, it SHOULD use set span status to `{process.executable.name}`. 

Instrumentations that have additional context about executed commands MAY use different span name format and SHOULD document it.
```

This would be a bit more prescriptive, would almost match what we do in databases + we span name should have low-ish cardinality according to the spec.

Author: I don't feel ? would as natural as in db here.  Something like this below?

```
The span name SHOULD be set to `{process.executable.name}`. 
Instrumentations that have additional context about executed commands MAY use a different low-cardinality span name format and SHOULD document it.
```

Reviewer 2: Looks good!
